### DISCUSSION TITLE
Reports in GridBase

### DISCUSSION THREAD

**Discussion by ArsenioInojosa** [dfaruque/Serenity.Extra#96] [Q&A] [answered]
Hello everyone,

Can you help me with an example of how to call the function to create the buttons to export to Excel, PDF and HTML.

How do I get it to enter the conditions and create the buttons?

`protected getButtons() {
        var buttons = super.getButtons();
        let reportRequest = this.getReportRequest();

        if (reportRequest.ListExcelServiceMethodName) {
            buttons.push(Ext.ExcelExportHelper.createToolButton({
                grid: this,
                service: this.getService() + '/' + reportRequest.ListExcelServiceMethodName,
                onViewSubmit: () => this.onViewSubmit(),
                separator: true
            }));
        }

        if (reportRequest.ReportKey) {
            buttons.push({
                title: q.text('Controls.ExportToPDF', 'Export to PDF'),
                icon: 'fa fa-file-pdf-o',
                onClick: () => {
                    let request = this.getReportRequest();
                    if (request)
                        Ext.ReportHelper.execute({ reportKey: reportRequest.ReportKey, params: { request: request } });
                }
            });
.....
        return buttons;
}`

The function protected getReportRequest(): ListReportRequest always returns this:

var request = Q.deepClone(view ? view.params : {}) 
request.ReportServiceMethodName = null;
request.ReportKey = null;
request.ListExcelServiceMethodName = q.ListExcelServiceMethodName;
request.EqualityFilterWithTextValue = {};
request.CustomParameters = {};

Values are never assigned.

Thank you

**Comment by dfaruque** [comment, marked as answer]
you should override `getReportRequest()` in your derived grid

```
        protected getReportRequest(): _Ext.ListReportRequest {
            var request = super.getReportRequest();
            request.ListExcelServiceMethodName = 'ListExcel';

            return request;
        }
```

**Comment by ArsenioInojosa** [comment]
Excellent thank you very much. Export to Excel is now working.

What value should I put in request.ReportKey?

**Reply by dfaruque** [reply]
```
    [Report("ns.SomeReport")] //this is the report key
    [ReportDesign(MVC.Views.....)]
    public class SomeReport : IReport
```
